source/solvers/fluid_dynamics_matrix_free_operators.cc

```diff
@@ -314,6 +314,62 @@ NavierStokesOperatorBase<dim, number>::compute_forcing_term()
     }
 }
 
+template <int dim, typename number>
+void
+NavierStokesOperatorBase<dim, number>::compute_buoyancy_term(
+  const VectorType                &temperature_solution,
+  const DoFHandler<dim>           &temperature_dof_handler,
+  const PhysicalPropertiesManager &physical_properties_manager)
+{
+  const auto thermal_expansion_model =
+    physical_properties_manager.get_thermal_expansion();
+
+  const double reference_temperature =
+    physical_properties_manager.get_reference_temperature();
+
+  const unsigned int n_cells =
+    matrix_free.n_cell_batches() + matrix_free.n_ghost_cell_batches();
+
+  FEValues<dim> fe_values(*(this->matrix_free.get_mapping_info().mapping),
+                          temperature_dof_handler.get_fe(),
+                          this->matrix_free.get_quadrature(),
+                          update_values);
+
+  buoyancy_term.reinit(n_cells, fe_values.n_quadrature_points);
+
+  std::vector<number> cell_temperature_solution(fe_values.n_quadrature_points);
+  std::vector<double> thermal_expansion(fe_values.n_quadrature_points);
+
+  for (unsigned int cell = 0; cell < n_cells; ++cell)
+    {
+      for (auto lane = 0u;
+           lane < matrix_free.n_active_entries_per_cell_batch(cell);
+           lane++)
+        {
+          fe_values.reinit(
+            matrix_free.get_cell_iterator(cell, lane)
+              ->as_dof_handler_iterator(temperature_dof_handler));
+
+          fe_values.get_function_values(temperature_solution,
+                                        cell_temperature_solution);
+
+          thermal_expansion_model->vector_value({}, thermal_expansion);
+
+          for (const auto q : fe_values.quadrature_point_indices())
+            {
+              const auto scaling =
+                1 - thermal_expansion[q] *
+                      (cell_temperature_solution[q] - reference_temperature);
```

Comment:
What is this  scaling term ? I am not sure I understand

Reply:
This is the first part of the buoyancy term (1 - β (T -Tref)).. it is simply precomputed before the following loop since only the quadrature points are needed. I have added the equations as comments for it to be more clear.

Reply:
Ohh ok so the scaling term is the scaling in front of the entire source term? Could you just add a comment to that matter?

Reply:
I see the new comment, great!